pkg/metrics/metrics.go

```diff
@@ -0,0 +1,126 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	MachineControllerFrozenDesc = prometheus.NewDesc("mcm_machine_controller_frozen", "Frozen status of the machine controller manager.", nil, nil)
+	MachineCountDesc            = prometheus.NewDesc("mcm_machine_items_total", "Count of machines currently managed by the mcm.", nil, nil)
+	MachineCreated              = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_created",
+		Help: "Creation time of the Machines currently managed by the mcm.",
+	}, []string{"name", "namespace", "uid"})
+
+	MachineCSPhase = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_current_status_phase",
+		Help: "Current status phase of the Machines currently managed by the mcm.",
+	}, []string{"name", "namespace", "uid", "phase"})
+
+	MachineInfo = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_info",
+		Help: "Information of the Machines currently managed by the mcm.",
+	}, []string{"name", "namespace", "uid", "generation", "kind", "api_version",
+		"spec_provider_id", "spec_class_api_group", "spec_class_kind", "spec_class_name"})
+
+	MachineStatusCondition = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_status_condition",
+		Help: "Information of the mcm managed Machines' status conditions.",
+	}, []string{"name", "namespace", "uid", "condition", "status"})
+
+	MachineSetCountDesc = prometheus.NewDesc("mcm_machineset_items_total", "Count of machinesets currently managed by the mcm.", nil, nil)
+	MachineSetCreated   = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_set_created",
+		Help: "Creation time of the Machinesets currently managed by the mcm.",
+	}, []string{"name", "namespace", "uid"})
+
+	MachineSetInfo = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_set_info",
+		Help: "Information of the Machinesets currently managed by the mcm.",
+	}, []string{"name", "namespace", "uid", "generation", "kind", "api_version", "spec_replicas",
+		"spec_min_ready_reconds", "spec_machine_class_api_group", "spec_machine_class_kind", "spec_machine_class_name"})
+
+	MachineSetStatusCondition = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_set_status_condition",
+		Help: "Information of the mcm managed Machinesets' status conditions.",
+	}, []string{"name", "namespace", "uid", "condition", "status"})
+
+	MachineSetStatusFailedMachines = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_set_failed_machines",
+		Help: "Information of the mcm managed Machinesets' failed machines.",
+	}, []string{"name", "namespace", "uid", "failed_Machine_name", "failed_machine_provider_id", "failed_machine_owner_ref",
+		"failed_Machine_last_operation_description", "failed_machine_last_operation_last_update_time", "failed_machine_last_operation_state",
```

Comment:
Please do not use labels to transport any kind of operation messages and try to avoid labels which can have many different values e.g. timestamps. In general labels should have a fixed value set, otherwise it would be very hard to query for those metrics in Prometheus.

Reply:
Removed these.